bot: only dismiss reviewers if we have 2+ internal approvals

zmb3 opened the pull request.

zmb3 committed:
bot: only dismiss reviewers if we have 2+ internal approvals

espadolini commented on .github/workflows/robot/internal/bot/check_test.go:
```diff
@@ -0,0 +1,141 @@
+					"user3": {},
+					"user4": {},
+				},
+				CodeReviewersOmit: make(map[string]bool),
+				DocsReviewers:     make(map[string]review.Reviewer),
+				DocsReviewersOmit: make(map[string]bool),
```
```suggestion
```

zmb3 replied:
Unfortunately required - `CheckAndSetDefaults` will fail otherwise.

espadolini commented on .github/workflows/robot/internal/bot/check.go:
```diff
@@ -128,9 +133,20 @@ func (b *Bot) reviewersToDismiss(ctx context.Context) ([]string, error) {
 	reviewedBy := make(map[string]struct{})
 	for _, review := range reviews {
 		reviewedBy[review.Author] = struct{}{}
+		if review.State == "APPROVED" && b.c.Review.IsInternal(review.Author) {
```
nit: replace the string literal with the `approved` const in `robot/internal/review`?

If we export those, same goes for the test cases in `check_test` I guess.

zmb3 replied:
Will do.

nklaassen commented on .github/workflows/robot/internal/bot/check.go:
```diff
@@ -128,9 +133,20 @@ func (b *Bot) reviewersToDismiss(ctx context.Context) ([]string, error) {
 	reviewedBy := make(map[string]struct{})
 	for _, review := range reviews {
 		reviewedBy[review.Author] = struct{}{}
+		if review.State == "APPROVED" && b.c.Review.IsInternal(review.Author) {
```
if the same author approves twice does it get double counted?

espadolini replied:
Reviews are sorted by time in github.go but they're not deduplicated by author, that's a good point.

We should reuse `review.reviewsByAuthor` here.

zmb3 replied:
Ughhh, great catch.

This seems unlikely, but worth fixing.

espadolini replied:
It would only be an issue if an admin approver approved twice

zmb3 replied:
Should be fixed either way now.

nklaassen replied:
and what happens now if an approver comments after approving (such as https://github.com/gravitational/teleport/pull/14434#pullrequestreview-1037897066)? We have a helper `reviewsByAuthor` which handles that case, could be useful?

nklaassen replied:
worst case we won't dismiss reviewers when we could have in this case, approving b/c that's better than what we have now

zmb3 replied:
Decided to leave this as is for now to fix the issue at hand.

Worst case, if someone comments after approval, we won't count it, and we may not dismiss reviewers when we could.

This is relatively uncommon, and if it does happen, we'll be no worse off than we are now. Having too many reviewers is better than having not enough.

zmb3 committed:
Only count each reviewer's latest review

This prevents an old approval from counting if the reviewer changes
their review state, and also ensures that multiple approvals from
the same author don't count as more than one approval.

zmb3 committed:
Replace review state magic strings with constants

zmb3 committed:
Merge branch 'master' into zmb3/fix-bot-dismissals

zmb3 committed:
Merge branch 'master' into zmb3/fix-bot-dismissals
